### PR TITLE
can now parse func params and types

### PR DESCRIPTION
**Pull request by colleend**
_Test Plan:_
Run:
``` semgrep-core -dump_ast tests/kotlin/types.kt ```

This gives me:
```
Pr(
  [DefStmt(
     ({
       name=EId(("foo", ()),
              {id_resolved=Ref(None); id_type=Ref(None);
               id_constness=Ref(None); });
       attrs=[]; tparams=[]; },
      FuncDef(
        {fkind=(Function, ());
         fparams=[ParamClassic(
                    {pname=Some(("x", ())); pdefault=None;
                     ptype=Some(TyQuestion((),
                                  TyTuple(
                                    [TyId(("String", ()),
                                       {id_resolved=Ref(None);
                                        id_type=Ref(None);
                                        id_constness=Ref(None); })])));
                     pattrs=[];
                     pinfo={id_resolved=Ref(Some((Param, 1)));
                            id_type=Ref(Some(TyQuestion((),
                                               TyTuple(
                                                 [TyId(("String", ()),
                                                    {id_resolved=Ref(
                                                     None);
                                                     id_type=Ref(None);
                                                     id_constness=Ref(
                                                     None); })]))));
                            id_constness=Ref(None); };
                     });
                  ParamClassic(
                    {pname=Some(("y", ())); pdefault=None;
                     ptype=Some(TyTuple(
                                  [TyId(("Int", ()),
                                     {id_resolved=Ref(None);
                                      id_type=Ref(None);
                                      id_constness=Ref(None); })]));
                     pattrs=[];
                     pinfo={id_resolved=Ref(Some((Param, 2)));
                            id_type=Ref(Some(TyTuple(
                                               [TyId(("Int", ()),
                                                  {id_resolved=Ref(None);
                                                   id_type=Ref(None);
                                                   id_constness=Ref(None); })])));
                            id_constness=Ref(None); };
                     });
                  ParamClassic(
                    {pname=Some(("z", ())); pdefault=None;
                     ptype=Some(TyTuple(
                                  [TyId(("Short", ()),
                                     {id_resolved=Ref(None);
                                      id_type=Ref(None);
                                      id_constness=Ref(None); })]));
                     pattrs=[];
                     pinfo={id_resolved=Ref(Some((Param, 3)));
                            id_type=Ref(Some(TyTuple(
                                               [TyId(("Short", ()),
                                                  {id_resolved=Ref(None);
                                                   id_type=Ref(None);
                                                   id_constness=Ref(None); })])));
                            id_constness=Ref(None); };
                     })];
         frettype=None;
         fbody=Block(
                 [ExprStmt(
                    Call(
                      Id(("x", ()),
                        {id_resolved=Ref(Some((Param, 1)));
                         id_type=Ref(Some(TyQuestion((),
                                            TyTuple(
                                              [TyId(("String", ()),
                                                 {id_resolved=Ref(None);
                                                  id_type=Ref(None);
                                                  id_constness=Ref(None); })]))));
                         id_constness=Ref(None); }),
                      [Arg(OtherExpr(OE_StmtExpr, [S(Return((), None, ()))]));
                       Arg(IdSpecial((This, ())))]), ())]);
         })))])
```

* params exist, although TyId id_resolved, id_type, id_constness is always None.
* added in support for modifiers (otherwise code would not compile), although not sure whether to use KeywordAttr Recursive for `Tail` or not. 
* added some initial support for annotations.

### CODE REVIEW

**Comment by aryx** [comment]
CI failures?

File: semgrep-core/parsing/Parse_kotlin_tree_sitter.ml [hunk 1]
```diff
@@ -73,10 +73,10 @@ let pat_b294348 (env : env) (tok : CST.pat_b294348) =
 
 let visibility_modifier (env : env) (x : CST.visibility_modifier) =
   (match x with
-   | `Public tok -> token env tok (* "public" *)
-   | `Priv tok -> token env tok (* "private" *)
-   | `Inte tok -> token env tok (* "internal" *)
-   | `Prot tok -> token env tok (* "protected" *)
+   | `Public tok -> KeywordAttr (Public, token env tok) (* "public" *)
+   | `Priv tok -> KeywordAttr (Private, token env tok) (* "private" *)
+   | `Inte tok -> NamedAttr (token env tok, [str env tok], empty_id_info (), fake_bracket []) (* "internal" *)
```
Comment (aryx) [inline semgrep-core/parsing/Parse_kotlin_tree_sitter.ml:78]: maybe factorize in an helper function like 'unhandled_keyword_attr_to_named_attr tok'

File: semgrep-core/parsing/Parse_kotlin_tree_sitter.ml [hunk 2]
```diff
@@ -153,16 +153,16 @@ let variance_modifier (env : env) (x : CST.variance_modifier) =
 
 let member_modifier (env : env) (x : CST.member_modifier) =
   (match x with
-   | `Over tok -> token env tok (* "override" *)
-   | `Late tok -> token env tok (* "lateinit" *)
+   | `Over tok -> NamedAttr (token env tok, [str env tok], empty_id_info (), fake_bracket []) (* "override" *)
```
Comment (aryx) [inline semgrep-core/parsing/Parse_kotlin_tree_sitter.ml:156]: I like this keyword. We should promote it in a new KeywordAttr constructor. OCaml also has this keyword.

File: semgrep-core/parsing/Parse_kotlin_tree_sitter.ml [hunk 3]
```diff
@@ -153,16 +153,16 @@ let variance_modifier (env : env) (x : CST.variance_modifier) =
...
+   | `Late tok -> NamedAttr (token env tok, [str env tok], empty_id_info (), fake_bracket []) (* "lateinit" *)
   )
 
 let class_modifier (env : env) (x : CST.class_modifier) =
   (match x with
-   | `Sealed tok -> token env tok (* "sealed" *)
-   | `Anno tok -> token env tok (* "annotation" *)
-   | `Data tok -> token env tok (* "data" *)
-   | `Inner tok -> token env tok (* "inner" *)
+   | `Sealed tok -> NamedAttr (token env tok, [str env tok], empty_id_info (), fake_bracket []) (* "sealed" *)
```
Comment (aryx) [inline semgrep-core/parsing/Parse_kotlin_tree_sitter.ml:162]: What is the difference between sealed and final in Java?
Maybe we can generate a Final here?

File: semgrep-core/parsing/Parse_kotlin_tree_sitter.ml [hunk 4]
```diff
@@ -216,12 +216,12 @@ let as_operator (env : env) (x : CST.as_operator) =
...
-   | `Op tok -> token env tok (* "operator" *)
-   | `Infix tok -> token env tok (* "infix" *)
-   | `Inline tok -> token env tok (* "inline" *)
-   | `Exte tok -> token env tok (* "external" *)
-   | `Susp tok -> token env tok (* "suspend" *)
+   | `Tail tok -> NamedAttr (token env tok, [str env tok], empty_id_info (), fake_bracket []) (* "tailrec" *)
+   | `Op tok -> NamedAttr (token env tok, [str env tok], empty_id_info (), fake_bracket []) (* "operator" *)
+   | `Infix tok -> NamedAttr (token env tok, [str env tok], empty_id_info (), fake_bracket []) (* "infix" *)
+   | `Inline tok -> NamedAttr (token env tok, [str env tok], empty_id_info (), fake_bracket []) (* "inline" *)
+   | `Exte tok -> NamedAttr (token env tok, [str env tok], empty_id_info (), fake_bracket []) (* "external" *)
```
Comment (aryx) [inline semgrep-core/parsing/Parse_kotlin_tree_sitter.ml:223]: You can use KeywordAttr Extern here.

File: semgrep-core/parsing/Parse_kotlin_tree_sitter.ml [hunk 5]
```diff
@@ -216,12 +216,12 @@ let as_operator (env : env) (x : CST.as_operator) =
...
-   | `Tail tok -> token env tok (* "tailrec" *)
-   | `Op tok -> token env tok (* "operator" *)
-   | `Infix tok -> token env tok (* "infix" *)
-   | `Inline tok -> token env tok (* "inline" *)
-   | `Exte tok -> token env tok (* "external" *)
-   | `Susp tok -> token env tok (* "suspend" *)
+   | `Tail tok -> NamedAttr (token env tok, [str env tok], empty_id_info (), fake_bracket []) (* "tailrec" *)
+   | `Op tok -> NamedAttr (token env tok, [str env tok], empty_id_info (), fake_bracket []) (* "operator" *)
+   | `Infix tok -> NamedAttr (token env tok, [str env tok], empty_id_info (), fake_bracket []) (* "infix" *)
+   | `Inline tok -> NamedAttr (token env tok, [str env tok], empty_id_info (), fake_bracket []) (* "inline" *)
```
Comment (aryx) [inline semgrep-core/parsing/Parse_kotlin_tree_sitter.ml:222]: This one also maybe we could promote to a regular KeywordAttr.

File: semgrep-core/parsing/Parse_kotlin_tree_sitter.ml [hunk 6]
```diff
@@ -1351,16 +1372,16 @@ and nullable_type (env : env) ((v1, v2) : CST.nullable_type) =
...
   let v3 = type_ env v3 in
-  let param = { (param_of_id v1) with ptype = Some v3 } in
-  ParamClassic param
+  v1, v3
 
 and parameter_modifiers (env : env) (x : CST.parameter_modifiers) =
   (match x with
-   | `Anno x -> annotation env x
+   | `Anno x -> let _ = annotation env x in
+       raise Todo
```
Comment (aryx) [inline semgrep-core/parsing/Parse_kotlin_tree_sitter.ml:1384]: why Todo? You can populate the pattrs field of parameter_classic no?

**Review by aryx — APPROVED** [review]

File: semgrep-core/parsing/Parse_kotlin_tree_sitter.ml (hunk 6 above)
Comment (colleend) [reply semgrep-core/parsing/Parse_kotlin_tree_sitter.ml:1384]: ah yes, I'll fix this

File: semgrep-core/parsing/Parse_kotlin_tree_sitter.ml (hunk 1 above)
Comment (colleend) [reply semgrep-core/parsing/Parse_kotlin_tree_sitter.ml:78]: good idea!

File: semgrep-core/parsing/Parse_kotlin_tree_sitter.ml (hunk 5 above)
Comment (colleend) [reply semgrep-core/parsing/Parse_kotlin_tree_sitter.ml:222]: I think thats a good idea 👍

File: semgrep-core/parsing/Parse_kotlin_tree_sitter.ml (hunk 3 above)
Comment (colleend) [reply semgrep-core/parsing/Parse_kotlin_tree_sitter.ml:162]: I think the sealed class is more of an enum class that you can have multiple instances of ("restricted class hierarchy where an object or a value can have one of the types from a limited set of values" from https://kotlinlang.org/), so its not the csharp sealed class 😓 

So sealed classes are different from final classes 😮

**Pull request merged**